### ISSUE TITLE
Watch starvation can cause OOMs

### ISSUE THREAD

**Issue by serathius** [etcd-io/etcd#16839]
### What would you like to be added?

Please read https://github.com/etcd-io/etcd/issues/16839#issuecomment-1989350422 for up to date information about the problem

**Comment by serathius** [comment]
Profiles I promised. They are less extreme than onces we have seen originally. They are more recent, as we reduced the write throughput later. Still they show the problem:

32% of allocations done by Range to get PrevKey for Watch

![image](https://github.com/etcd-io/etcd/assets/5873433/3d97b39e-27dd-46ae-88e8-cf08861a353b)

57% of cpu used on GC
![image](https://github.com/etcd-io/etcd/assets/5873433/170de964-c7ce-4386-9cc3-f90ab94a304d)

**Comment by serathius** [comment]
One clarification, access to prevKey in https://github.com/etcd-io/etcd/blob/a4f507c30a7b1335a240b9b9c263501c3d7262a1/server/storage/mvcc/kvstore_txn.go#L203 doesn't touch bbolt, only the in memory index.

**Comment by chaochn47** [comment]
xref. https://github.com/etcd-io/etcd/issues/17529#issuecomment-1980258283.

**Comment by chaochn47** [comment]
The bigger issue about `prevKV` is for a single write, we copy the event for the matching watchers and for each watcher, we copy the entire transaction buffer just to get prevKV and it would duplicate 100 time (let's say there are 100 watchers). 

Hence it is not efficient IMHO.

**Comment by serathius** [comment]
Looked into performance of prevKV, it is abysmal, don't know how we released such feature and how Kubernetes adopted it without noticing. 

I have modified the [watch latency benchmark](https://github.com/etcd-io/etcd/blob/main/tools/benchmark/cmd/watch_latency.go) to be useful (will share code soon). 

For 100KB objects, 100 watchers, 100 put qps (similar to setup in https://github.com/kubernetes/kubernetes/issues/123448) I got:

<google-sheets-html-origin></style>
prevKV | etcd memory [GB] | etcd cpu [core] | 90%ile event latency[s] | 99%ile watch event latency[s] | 99.9%ile watch event latency[s]
-- | -- | -- | -- | -- | --
disabled | 1.1 | 0.76 | 0.2s | 0.73s | 1s
enabled | 8.6 | 1 | 16s | 18s | 25.7s

The @chaochn47 fix for compacted watchers will not help here, as there was no compaction done here. We need to address it even if it's a performance change.

**Comment by serathius** [comment]
Deleted the comment above because is was totally inaccurate.

tldr; Issue is not related to prevKV, but to overall watch efficiency during watch starvation

The main reason that enabling prevKV had such impact is fact that it doubles the event size, meaning it puts the twice as much constrain on the response stream. In my testing I was close to the response throughput limit even without prevKV, so enabling it caused the memory leak. I repeated the tests where I compared watch streams based on the response throughput (measured using `etcd_network_client_grpc_sent_bytes_total` metric).

In the repeated tests I have found that prevKV is totally not at fault, the issues is caused by inefficient behavior of `syncWatchers` when the watch stream is clogged. It will repeatably try to sync watchers every 100ms rereading all the events, but effectively syncing only small subset of watchers. In my testing with 100 watchers, I have observed that sync loop touched each revision 10 times, meaning we allocate 10 times more memory than needed. I also found that 
for 100 watchers, reading prevKV 100 times per event is less costly than rereading event by `syncWatchers`.

<google-sheets-html-origin>
prevKV | object size | response throughput[MB/s] | etcd memory [GB] | 90%ile watch event latency[s] | 99%ile watch event latency[s] | 99.9%ile watch event latency[s] | 90%ile put latency[s] | 99%ile put latency[s] | 99.9%ile put latency[s]
-- | -- | -- | -- | -- | -- | -- | -- | -- | --
TRUE | 50KB | 1003 | 0.389 | 0.116 | 0.57 | 0.843 | 0.001 | 0.003 | 0.01
FALSE | 100KB | 1065 | 0.5637 | 0.099 | 0.495 | 0.6629 | 0.001 | 0.004 | 0.012
TRUE | 100KB | 1591 | 4 | 6.56 | 10.16 | 12.78 | 0.002 | 0.007 | 0.014
FALSE | 200KB | 1670 | 6.1 | 5.5 | 8.34 | 9.96 | 0.003 | 0.011 | 0.023

**Comment by serathius** [comment]
Amazing just reducing the syncWatchers frequency improves watch latency and memory usage.

<google-sheets-html-origin>
sync duration[s] | response throughput[MB/s] | etcd memory [GB] | etcd cpu [core] | 90%ile watch event latency[s] | 99%ile watch event latency[s]
-- | -- | -- | -- | -- | --
0.1 (current) | 1620 | 6.6 | 6.34 | 9.64 | 14.04
0.5 | 1671 | 3.2 | 5.46 | 8.4 | 11.04
1 (proposed) | 1736 | 2.1 | 4.29 | 6.18 | 7.78
2 | 1728 | 1.7 | 5.19 | 6.53 | 7.44
5 | 1204 | 2.2 | 9.38 | 13.42 | 16.43

Looks like the improvements peek around 1s, improving the watch response throughput by 4%, reducing memory allocations by 4-5 times (base memory for puts is 850MB) and reducing the latency 30-45%. This already looks like great improvement.

A proper implementation of reusing events will be complicated, making backport questionable, thus I propose to a simple backportable solution to reduce impact of slow watchers.

**Proposal**: Reduce the syncLoop period from 0.1s to 1s

Want to confirm how the original value was proposed.

Note; there is no risk of syncLoop not catching up with unsynched watchers as it also includes shortcut if it cannot keep up https://github.com/etcd-io/etcd/blob/ddf54715bf47b8c7ec2f1257b25cd0981d081ae9/server/storage/mvcc/watchable_store.go#L213-L245

**Comment by serathius** [comment]
Validated that 1s performs better than 0.1s for different number of streams (20-10000):

<google-sheets-html-origin>
object size | streams | sync duration[s] | response throughput[MB/s] | etcd memory [GB] | 50%ile watch event latency[s] | 90%ile watch event latency[s] | 99%ile watch event latency[s] | 99.9%ile watch event latency[s]
-- | -- | -- | -- | -- | -- | -- | -- | --
1MB | 20 | 0.1 | 1579 | 14.9 | 3.47 | 6.32 | 8.22 | 9.4
1MB | 20 | 1 | 1634 | 9 | 3.45 | 5.63 | 7.98 | 9.34
30KB | 1000 | 0.1 | 2006 | 5.5 | 4.26 | 12.74 | 18.81 | 22.69
30KB | 1000 | 1 | 2257 | 1.4 | 4.07 | 11.76 | 17.8 | 21.87
3KB | 10000 | 0.1 | 2061 | 1.8 | 4.78 | 12.82 | 19.33 | 23.68
3KB | 10000 | 1 | 2154 | 1.1 | 4.29 | 11.74 | 18.01 | 22.36

**Comment by serathius** [comment]
synchWatchers period was always set to 100ms from first implementation of watch https://github.com/etcd-io/etcd/pull/3507 Didn't find any discussion pointing to why 100ms was picked.

This makes me feel convinced that we can change it.

**Comment by ahrtr** [comment]
Thanks for the performance test results.

> **Proposal**: Reduce the syncLoop period from 0.1s to 1s

I think it's more prudent to introduce a config item instead of hard coding the sync loop interval/frequency.

Since the history will never change (of course it can be compacted), so the idea of reusing events (https://github.com/etcd-io/etcd/pull/17563) should be OK.

We should also consider to limit the max data/events to read. Currently it always read all the data/events in range [minRev, curRev]. Obviously it may consume huge memory if there are huge number of events in the range.
https://github.com/etcd-io/etcd/blob/d639abe1aa1d357cd1f58ee313c6df50029c912f/server/storage/mvcc/watchable_store.go#L363

**Comment by serathius** [comment]
Reducing syncLoop causes issue with resynchronization of new watch requests opened on concrete revision. Instead I'm proposing the two improvements:
* https://github.com/etcd-io/etcd/pull/17563 
* https://github.com/etcd-io/etcd/pull/18975
